_posts/2020-01-13-serverless-express-api.md

```diff
@@ -0,0 +1,562 @@
+    Properties:
+      CodeUri: hello-world/
+      Handler: app.lambdaHandler
+      Runtime: nodejs12.x
+      Events:
+        HelloWorld:
+          Type: HttpApi
+          Properties:
+            Path: /{proxy+}
+            Method: get
+
+Outputs:
+  HelloWorldApi:
+    Description: API Gateway endpoint URL for Prod stage for Hello World function
+    Value:
+      Fn::Sub: https://${ServerlessHttpApi}.execute-api.${AWS::Region}.amazonaws.com/
+  ExpressBackendFunction:
+    Description: Express Backend Lambda Function ARN
+    Value: !Sub ExpressBackend.Arn
+  ExpressBackendIamRole:
+    Description: Implicit IAM Role created for Hello World function
+    Value: !Sub ExpressBackendFunctionRole.Arn
+
+```
+
+We don't really need to change much. All that needed to be done was 
+- modifying the event type to `HttpApi`
+- change the path to `/{proxy}+`. This will ensure all get requests are passed to the lambda request, regardless of their path.
+- Change the output to reference `${ServerlessHttpApi}` instead of `${ServerlessRestApi}`
+
+# Build, Test, & Deploy the App
+
+Firstly, we execute `sam build` from our base directory. This should result in the following output from the sam-cli.
+
+```bash
+Building resource 'ExpressBackend'
+Running NodejsNpmBuilder:NpmPack
+Running NodejsNpmBuilder:CopyNpmrc
+Running NodejsNpmBuilder:CopySource
+Running NodejsNpmBuilder:NpmInstall
+Running NodejsNpmBuilder:CleanUpNpmrc
+
+Build Succeeded
+
+Built Artifacts  : .aws-sam/build
+Built Template   : .aws-sam/build/template.yaml
+
+Commands you can use next
+=========================
+[*] Invoke Function: sam local invoke
+[*] Deploy: sam deploy --guided
+```
+
+Assuming that you have docker installed, you can use `sam local invoke` as a quick test. If you do, your output will look like this;
+
+```bash
+▶ sam local invoke
+Invoking app.lambdaHandler (nodejs12.x)
+
+Fetching lambci/lambda:nodejs12.x Docker container image......
+Mounting /Users/matt.tyler/projects/simple-node-api/simple-node-api/.aws-sam/build/ExpressBackend as /var/task:ro,delegated inside runtime container
+START RequestId: 6bb44d66-e096-124b-5ce9-5f1f1fea88f9 Version: $LATEST
+2020-01-02T06:00:30.213Z        6bb44d66-e096-124b-5ce9-5f1f1fea88f9    ERROR   (node:17) [DEP0066] DeprecationWarning: OutgoingMessage.prototype._headers is deprecated
+END RequestId: 6bb44d66-e096-124b-5ce9-5f1f1fea88f9
+REPORT RequestId: 6bb44d66-e096-124b-5ce9-5f1f1fea88f9  Init Duration: 473.40 ms        Duration: 10.32 ms  Billed Duration: 100 ms  Memory Size: 128 MB     Max Memory Used: 50 MB
+
+{"statusCode":200,"headers":{"x-powered-by":"Express","content-type":"text/html; charset=utf-8","content-length":"11","etag":"W/\"b-Ck1VqNd45QIvq3AZd8XYQLvEhtA\""},"isBase64Encoded":false,"body":"Hello World"}
+```
+
+We can see the 'hello world' response inside the JSON payload that was returned from our lambda function.
+
+Now let's deploy the application - to do this we will perform a guided deploy.
```

Comment:
why a guided deploy? may be a stupid question, I'm just interested..

Reply:
Upon completion of a guided deploy, a configuration file will be created that allows future deploys to use our previous responses.

mine looked like this;

```toml
version = 0.1
[default]
[default.deploy]
[default.deploy.parameters]
stack_name = "simple-node-api"
s3_bucket = "aws-sam-cli-managed-default-samclisourcebucket-5elb445a2eu1"
s3_prefix = "simple-node-api"
region = "ap-southeast-2"
confirm_changeset = true
capabilities = "CAPABILITY_IAM"
```

It's a new feature that only got released around re:invent.